handlers/firebase/handler.go

```diff
@@ -197,19 +221,83 @@ func (h *handler) Send(ctx context.Context, msg courier.MsgOut, res *courier.Sen
 			return courier.ErrResponseStatus
 		}
 
-		// was this successful
-		success, _ := jsonparser.GetInt(respBody, "success")
-		if success != 1 {
+		responseName, err := jsonparser.GetString(respBody, "name")
+		if err != nil {
 			return courier.ErrResponseUnexpected
 		}
 
-		externalID, err := jsonparser.GetInt(respBody, "multicast_id")
-		if err != nil {
+		if !strings.Contains(responseName, fmt.Sprintf("projects/%s/messages/", fcmAuthJSON["project_id"])) {
 			return courier.ErrResponseUnexpected
 		}
-		res.AddExternalID(fmt.Sprintf("%d", externalID))
+		externalID := strings.TrimLeft(responseName, fmt.Sprintf("projects/%s/messages/", fcmAuthJSON["project_id"]))
+		if externalID == "" {
+			return courier.ErrResponseUnexpected
+		}
+
+		res.AddExternalID(externalID)
 
 	}
 
 	return nil
 }
+
+func (h *handler) getAccessToken(ctx context.Context, channel courier.Channel, clog *courier.ChannelLog) (string, error) {
+	rc := h.Backend().RedisPool().Get()
+	defer rc.Close()
+
+	tokenKey := fmt.Sprintf("channel-token:%s", channel.UUID())
+
+	h.fetchTokenMutex.Lock()
+	defer h.fetchTokenMutex.Unlock()
+
+	token, err := redis.String(rc.Do("GET", tokenKey))
+	if err != nil && err != redis.ErrNil {
+		return "", fmt.Errorf("error reading cached access token: %w", err)
+	}
+
+	if token != "" {
+		return token, nil
+	}
+
+	token, expires, err := h.fetchAccessToken(ctx, channel, clog)
+	if err != nil {
+		return "", fmt.Errorf("error fetching new access token: %w", err)
+	}
+
+	_, err = rc.Do("SET", tokenKey, token, "EX", int(expires/time.Second))
+	if err != nil {
+		return "", fmt.Errorf("error updating cached access token: %w", err)
+	}
+
+	return token, nil
+}
+
+// fetchAccessToken tries to fetch a new token for our channel, setting the result in redis
+func (h *handler) fetchAccessToken(ctx context.Context, channel courier.Channel, clog *courier.ChannelLog) (string, time.Duration, error) {
+
+	fcmAuthJSONString := channel.StringConfigForKey(configAuthJSON, "")
+	if fcmAuthJSONString == "" {
+		return "", 0, courier.ErrChannelConfig
+	}
+
+	var fcmAuthJSON map[string]string
+
+	err := json.Unmarshal([]byte(fcmAuthJSONString), &fcmAuthJSON)
+	if err != nil {
+		return "", 0, courier.ErrChannelConfig
+	}
+
+	sendURL := fmt.Sprintf("https://fcm.googleapis.com/v1/projects/%s/messages:send", fcmAuthJSON["project_id"])
+
+	ts, err := idtoken.NewTokenSource(ctx, sendURL, option.WithCredentialsJSON([]byte(fcmAuthJSONString)))
```

Comment:
Use the credentials JSON to make a token source and we use REST API for the requests with the Bearer access token
https://cloud.google.com/docs/authentication/get-id-token#go